docs/fig.1
'\" t
.\"     Title: fig
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 05/28/2017
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "FIG" "1" "05/28/2017" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
fig \- Manage and deploy configuration profiles\&.
.SH "SYNOPSIS"
.sp
.nf
\fIfig add\fR <app> [\-\-url=<url>]
\fIfig add\fR <app>/<profile> [\-\-extend=<parent>]
\fIfig deploy\fR <app>/<profile>
\fIfig edit\fR <app>/<profile>
\fIfig remove\fR <app>[/<profile>]
\fIfig show\fR [\-a] [\-l] [<app>]
\fIfig snapshot\fR <app>/<profile>
.fi
.SH "DESCRIPTION"
.sp
Fig helps you create, manage, and deploy profiles to quickly bring application configuration files and any other assets to a known state\&.
.SH "OPTIONS"
.PP
\-\-version
.RS 4
Prints the current
\fIfig\fR
version\&.
.RE
.SH "COMMANDS"
.PP
\fIadd\fR
.RS 4
Creates new top\-level app <app> or adds a configuration profile <profile> to an existing app\&.
.sp
When a profile is added using \-\-extend=<parent>, Fig creates the profile with a default action "extend: <parent>"\&.
.sp
When an app is created with
\-\-url=<url>, Fig attempts to clone the Git repository at <url>\&.
.sp
NOTE \(em Creating Fig profiles from Git repositories is not supported\&.
.RE
.PP
\fIdeploy\fR
.RS 4
Loads <app>/<profile>\*(Aqs YAML definition and executes each action sequentially\&.
.sp
If <app>/<profile>
\fIextends\fR
another Fig profile, any
\fIfile\fR
actions will use the extending profile\(cqs assets\&.
.RE
.PP
\fIedit\fR
.RS 4
Opens <app>/<profile>\*(Aqs YAML file in $EDITOR\&.
.RE
.PP
\fIremove\fR
.RS 4
Deletes top\-level apps and all their profiles, or just individual profiles\&.
.RE
.PP
\fIshow\fR
.RS 4
Lists all local Fig apps, or a single app\(cqs profiles\&.
.sp
Use option
\-a
to list hidden apps or profiles (named with a leading "\&.")\&. Use option
\-l
to list in long format\&.
.RE
.PP
\fIsnapshot\fR
.RS 4
If <app>/<profile> contains
\fIfile:replace\fR
actions, Fig will update profile assets using their live counterparts, copying from
\fIpath\fR
to
\fIsource\fR\&.
.RE


CHANGELOG.md
# Change Log

All notable changes to Fig will be documented in this file.

## Unreleased

## [0.4.1] - 2017-06-04
### Added
- `file:replace_string`
- [#19](https://github.com/ashur/fig/issues/19) – Add `--extend` option to `add` command

### Fixed
- [#17](https://github.com/ashur/fig/issues/17) – `show` no longer highlights extending profiles
- [#18](https://github.com/ashur/fig/issues/18) – Invalid YAML is not correctly handled

## [0.4.0] - 2017-05-23
### Added
- Support for nested variable definitions in multi-level `include`s

### Changed
- Lazy-load profile definition files

### Fixed
- Validation of profile and action definitions
- Prevent profile from including itself

### Removed
- `action` and `path`-less File action syntax
- `install` command
- `upgrade` command

## [0.3.1] - 2017-05-05
### Fixed
- Always setting `ignore_errors`, `ignore_output` to true
- Variable replacement in `file:source` values
- Handling of unexpected YAML values

### Removed
- Experimental privilege escalation support

## [0.3.0] - 2017-01-02
### Added
- `install` command
- man page, fig.1
- `upgrade` command
- Expanded `show` options
- `action` and `path` properties to File action syntax

### Changed
- Renamed `update` to `snapshot`

### Deprecated
- `action` and `path`-less File action syntax

### Fixed
- Skip asset deletion for profiles with no assets

### Changed
- Switched to Cranberry\CLI framework

## [0.2.1] - 2016-12-21
### Added
- Support for creating Fig apps by cloning Git repositories

## [0.2.0] - 2016-12-17
### Added
- Support for profile variables

## [0.1.1] - 2016-12-13
### Fixed
- Sanitize command strings


lib/src/Fig/Fig.php
<?php

/*
 * This file is part of Fig
 */
namespace Fig;

use Cranberry\CLI\Format;
use Cranberry\CLI\Shell;
use Cranberry\Core\File;
use Spyc;

class Fig
{
	const STRING_INVALID_PROPERTY_VALUE = "Invalid value for property '%s': %s";
	const STRING_MISSING_REQUIRED_PROPERTY = "Missing required property '%s'";

	/**
	 * @var	array
	 */
	protected $appDirs=[];

	/**
	 * @var	array
	 */
	protected $apps=[];

	/**
	 * @var	Huxtable\Core\File\Directory
	 */
	protected $figDirectory;

	/**
	 * @return	void
	 */
	public function __construct( File\Directory $figDirectory )
	{
		$this->figDirectory = $figDirectory;

		if( !$this->figDirectory->exists() )
		{
			$this->figDirectory->create();
		}

		// Only include directories
		$fileFilter = new File\Filter();
		$fileFilter->setDefaultMethod( File\Filter::METHOD_INCLUDE );
		$fileFilter->addInclusionRule( function( $file )
		{
			return $file->isDir();
		});

		$apps = $this->figDirectory->children( $fileFilter );

		foreach( $apps as $appDir )
		{
			// Add Directory object for later instantiation
			$appName = $appDir->getBasename();
			$this->appDirs[$appName] = $appDir;
		}
	}

	/**
	 * @param	Fig\App		$app
	 * @return	self
	 */
	public function addApp( App $app )
	{
		$this->apps[$app->getName()] = $app;
		return $this;
	}

	/**
	 * Create an template app folder
	 *
	 * @param	string	$newAppName
	 * @return	Fig\App
	 */
	public function createApp( $newAppName )
	{
		/* Create new instance to validate $newAppName value */
		$app = new App( $newAppName );
		$appName = $app->getName();

		$appDir = $this->figDirectory->childDir( $appName );

		if( $appDir->exists() )
		{
			throw new \Exception( "App '{$appName}' already exists. See 'fig show'." );
		}

		/* Create directories on disk */
		$appDir->create();

		$assetsDir = $appDir->childDir( Profile::ASSETS_DIRNAME );
		$assetsDir->create();

		/* Update the internal inventory, just in case */
		$this->appDirs[$appName] = $appDir;
		ksort( $this->appDirs );

		$app = $this->getApp( $appName );
		return $app;
	}

	/**
	 * @param	string	$appName
	 * @param	string	$repoURL
	 * @return	void
	 */
	public function createAppFromRepository( $appName, $repoURL )
	{
		$appDir = $this->figDirectory->childDir( $appName );
		$commandClone = sprintf( 'git clone %s %s', $repoURL, $appDir );

		echo 'Cloning... ';
		$result = Shell::exec( $commandClone, true, '   > ' );

		if( $result['exitCode'] == 0 )
		{
			echo 'done.' . PHP_EOL;
			return true;
		}

		echo 'failed:' . PHP_EOL . PHP_EOL;
		echo $result['output']['formatted'] . PHP_EOL;
		return false;
	}

	/**
	 * @param	string	$appName
	 * @param	string	$profileName
	 * @param	string	$actionContents
	 * @return	void
	 */
	public function createProfile( $appName, $profileName, $actionContents )
	{
		if( !isset( $this->appDirs[$appName] ) )
		{
			throw new \OutOfRangeException( "App not found '{$appName}'" );
		}

		$appDir = $this->appDirs[$appName];

		/* Create the assets directory */
		$assetsDir = $appDir
			->childDir( Profile::ASSETS_DIRNAME )
			->childDir( $profileName );
		$assetsDir->create();

		$profileFile = $appDir->child( "{$profileName}.yml" );
		if( $profileFile->exists() )
		{
			throw new \Exception( "Profile '{$appName}/{$profileName}' already exists. See 'fig show'." );
		}

		$profileContents = <<<PROFILE
# {$profileName}
---

{$actionContents}

PROFILE;

		/* Update the internal inventory */
		$profileFile->putContents( $profileContents );

		/* Create a stub object */
		if( isset( $this->apps[$appName] ) )
		{
			$profile = new Profile( $profileName, $appName );
			$this->apps[$appName]->addProfile( $profile );
		}
	}

	/**
	 * Decode data with a consistent format (currently YAML)
	 *
	 * @param	Huxtable\Core\File\File	$file
	 * @return	array
	 */
	static public function decodeFile( File\File $file )
	{
		$fileContents = $file->getContents();
		$data = Spyc::YAMLLoadString( $fileContents );
		return $data;
	}

	/**
	 * Decode data with a consistent format (currently YAML)
	 *
	 * @param	string	$string
	 * @return	array
	 */
	static public function decodeString( $string )
	{
		$data = Spyc::YAMLLoadString( $string );
		return $data;
	}

	/**
	 * @param	string	$appName
	 * @return	void
	 */
	public function deleteApp( $appName )
	{
		$appDir = $this->appDirs[$appName];

		/*
		 * Delete the source directory
		 *
		 * Note: This uses a shell command instead of the Directory::delete method
		 *       because `-f` is not available via PHP filesystem functions.
		 *       Since apps are likely under version control, `-f` is required
		 *       to remove the app directory without user intervention.
		 */
		$command = "rm -rf '{$appDir}'";
		$result = Shell::exec( $command );
		if( $result['exitCode'] !== 0 )
		{
			throw new \Exception( $result['output']['raw'], $result['exitCode'] );
		}

		/* Update the internal inventory */
		unset( $this->appDirs[$appName] );

		if( isset( $this->apps[$appName] ) )
		{
			unset( $this->apps[$appName] );
		}
	}

	/**
	 * @param	string	$appName
	 * @param	string	$profileName
	 * @return	void
	 */
	public function deleteProfile( $appName, $profileName )
	{
		$appDir = $this->appDirs[$appName];

		/* Delete the assets directory */
		$assetsDir = $appDir
			->childDir( Profile::ASSETS_DIRNAME )
			->childDir( $profileName );

		if( $assetsDir->exists() )
		{
			$assetsDir->delete();
		}

		/* Delete the source file */
		$profileFile = $appDir->child( "{$profileName}.yml" );

		if( $profileFile->exists() )
		{
			$profileFile->delete();
		}

		/* Update the internal inventory */
		if( isset( $this->apps[$appName] ) )
		{
			$this->apps[$appName]->removeProfile( $profileName );
		}
	}

	/**
	 * @param	string	$appName
	 * @param	string	$profileName
	 * @param	array	$variables
	 * @return	void
	 */
	public function deployProfile( $appName, $profileName, array $variables=null )
	{
		$app = $this->getApp( $appName );

		/*
		 * Actions
		 */
		$actions = $app->getProfileActions( $profileName );
		foreach( $actions as $action )
		{
			/* Variables */
			if( !is_null( $variables ) )
			{
				$action->setVariables( $variables );
			}

			$outputColor = 'green';
			$actionTitle = $action->getTitle();

			/* Set Fig directory for 'file' actions */
			if( $action->usesFigDirectory )
			{
				$action->setFigDirectory( $this->figDirectory );
			}

			self::outputActionTitle( $action->type, $actionTitle );

			try
			{
				$result = $action->execute();
			}
			catch( \Exception $e )
			{
				$result['error'] = true;
				$result['output'] = $e->getMessage();
			}
			$output = empty( $result['output'] ) ? 'OK' : $result['output'];

			if( $result['error']  )
			{
				$outputColor = 'red';
			}

			if( !isset( $result['silent'] ) || !$result['silent'] )
			{
				self::outputAction( $output, $outputColor );
			}

			if( $action->usesDeprecatedSyntax )
			{
				$deprecationWarning = new Format\String( 'WARNING: This action uses a deprecated syntax. See https://github.com/ashur/fig/wiki/Actions' );
				$deprecationWarning->foregroundColor( 'yellow' );

				echo $deprecationWarning . PHP_EOL;
			}
		}
	}

	/**
	 * Encode data with a consistent format (currently YAML)
	 *
	 * @param	array	$data
	 * @return	string
	 */
	static public function encodeData( array $data )
	{
		$encoded = Spyc::YAMLDump( $data, 4, 0 );
		return $encoded;
	}

	/**
	 * @param	string	$appName
	 * @return	Fig\App
	 */
	public function getApp( $appName )
	{
		if( !isset( $this->apps[$appName] ) )
		{
			if( !isset( $this->appDirs[$appName] ) )
			{
				throw new \OutOfRangeException( "App '{$appName}' not found." );
			}

			// Lazy load the app
			$app = App::getInstanceFromDirectory( $this->appDirs[$appName] );
			$this->addApp( $app );
		}

		return $this->apps[$appName];
	}

	/**
	 * @return	array
	 */
	public function getApps()
	{
		$apps = $this->apps;

		foreach( $this->appDirs as $appName => $appDir )
		{
			if( !isset( $apps[$appName] ) )
			{
				$apps[$appName] = App::getInstanceFromDirectory( $appDir );
			}
		}

		return array_values( $apps );
	}

	/**
	 * @return	Cranberry\Core\File\Directory
	 */
	public function getFigDirectory()
	{
		return $this->figDirectory;
	}

	/**
	 * Convert user-supplied values to strings for output
	 *
	 * @param	mixed	$value
	 * @return	string
	 */
	public static function getStringRepresentation( $value )
	{
		$stringValue = json_encode( $value, true );
		if( json_last_error() != JSON_ERROR_NONE )
		{
			$stringValue = var_export( $value, true );
		}

		return $stringValue;
	}

	/**
	 * @param	string	$string
	 * @param	array	$variables
	 * @return	void
	 */
	static public function replaceVariables( $string, array $variables )
	{
		foreach( $variables as $key => $value )
		{
			$string = str_replace( "{{{$key}}}", $value, $string );
			$string = str_replace( "{{ {$key} }}", $value, $string );
		}

		return $string;
	}

	/**
	 * @param	mixed	$output
	 * @param	string	$outputColor
	 * @return	void
	 */
	public function outputAction( $output, $outputColor )
	{
		$outputString = new Format\String();
		$outputString->foregroundColor( $outputColor );

		if( is_scalar( $output ) )
		{
			$output = [$output];
		}

		foreach( $output as $line )
		{
			if( $line === false )
			{
				$line = 'false';
			}
			if( $line === true )
			{
				$line = 'true';
			}

			$outputString->setString( $line );
			echo $outputString . PHP_EOL;
		}
	}

	/**
	 * @param	string	$category
	 * @param	string	$actionTitle
	 */
	public function outputActionTitle( $category, $title )
	{
		if( is_null( $title ) )
		{
			return;
		}

		echo PHP_EOL;

		$category = strtoupper( $category );
		echo sprintf( "%'*-80s", "{$category}: {$title} " ) . PHP_EOL;
	}

	/**
	 * Update a profile's source files using their targets
	 *
	 * @param	Fig\Profile	$profile
	 */
	public function updateProfileAssetsFromTarget( Profile $profile )
	{
		$profile->updateAssetsFromTarget( $this->figDirectory );
	}
}
